starboard/shared/libiamf/iamf_buffer_parser.h

```diff
@@ -0,0 +1,109 @@
+#include "starboard/shared/starboard/player/input_buffer_internal.h"
+
+namespace starboard {
+namespace shared {
+namespace libiamf {
+
+class BufferReader;
+
+// TODO: Skip parsing if the OBUs are redundant
+class IamfBufferParser {
+ public:
+  typedef ::starboard::shared::starboard::player::InputBuffer InputBuffer;
+
+  struct IamfBufferInfo {
+    bool is_valid() const {
+      return mix_presentation_id.has_value() && sample_rate > 0 &&
+             num_samples > 0;
+    }
+
+    uint32_t num_samples;
+    int sample_rate;
+    std::optional<uint32_t> mix_presentation_id;
+    std::vector<uint8_t> config_obus;
+    size_t config_obus_size;
+    std::vector<uint8_t> data;
+    size_t data_size;
+    const scoped_refptr<InputBuffer> input_buffer;
+  };
+
+  IamfBufferParser();
+
+  bool ParseInputBuffer(const scoped_refptr<InputBuffer>& input_buffer,
+                        IamfBufferInfo* info,
+                        const bool prefer_binaural_audio,
+                        const bool prefer_surround_audio);
+
+ private:
+  // Used in the selection of a binaural mix presentation, using the strategy
+  // defined in
+  // https://aomediacodec.github.io/iamf/#processing-mixpresentation-selection.
+  // The preferred methods of choosing a binaural mix presentation are listed
+  // from high to low.
+  enum BinauralMixSelection {
+    kBinauralMixSelectionLoudspeakerLayout,
+    kBinauralMixSelectionLoudnessLayout,
+    kBinauralMixSelectionNotFound
+  };
+
+  bool ParseInputBufferInternal(const scoped_refptr<InputBuffer>& input_buffer,
+                                IamfBufferInfo* info,
+                                const bool prefer_binaural_audio,
+                                const bool prefer_surround_audio);
+  // Reads a single Descriptor OBU. Returns false on error.
+  bool ParseDescriptorOBU(BufferReader* reader,
+                          IamfBufferInfo* info,
+                          const bool prefer_binaural_audio,
+                          const bool prefer_surround_audio);
+  bool ParseOBUHeader(BufferReader* reader,
+                      uint8_t* obu_type,
+                      uint32_t* obu_size) const;
+  bool ParseCodecConfigOBU(BufferReader* reader, IamfBufferInfo* info);
+  bool ParseAudioElementOBU(BufferReader* reader,
+                            IamfBufferInfo* info,
+                            const bool prefer_binaural_audio,
+                            const bool prefer_surround_audio);
+  bool ParseMixPresentationOBU(BufferReader* reader,
+                               IamfBufferInfo* info,
+                               const bool prefer_binaural_audio,
+                               const bool prefer_surround_audio);
+  // Helper function to skip parsing ParamDefinitions found in the config OBUs
+  // https://aomediacodec.github.io/iamf/v1.0.0-errata.html#paramdefinition
+  bool SkipParamDefinition(BufferReader* reader) const;
+
+  std::unordered_set<uint32_t> binaural_audio_element_ids_;
```

Comment:
We can store audio element ids as a local variablesin ParseInputBufferInternal() instead of a instance variable.